### PR TITLE
renderer: Add upscaling support

### PR DESCRIPTION
**Pull request by Macdu**
Add upscaling support to Vita3K.

The most important thing left to do is to implement surface readback to the cpu. It isn't hard but I need first to fix and merge #1770 in order for it not to kill performance.

### CODE REVIEW

File: vita3k/renderer/src/gl/surface_cache.cpp [hunk 1]
```diff
@@ -58,18 +58,24 @@ void GLSurfaceCache::do_typeless_copy(const GLint dest_texture, const GLint sour
...
     const SceGxmColorBaseFormat base_format, Ptr<void> address, SurfaceTextureRetrievePurpose purpose, std::uint32_t &swizzle,
     std::uint16_t *stored_height, std::uint16_t *stored_width) {
     // Create the key to access the cache struct
     const std::uint64_t key = address.address();
 
+    const uint32_t original_width = width;
+    const uint32_t original_height = height;
+
+    width *= state.res_multiplier;
+    height = height *= state.res_multiplier;
```
Comment (Zangetsu38) [inline vita3k/renderer/src/gl/surface_cache.cpp:71]:
```suggestion
    height *= state.res_multiplier;
```

File: vita3k/gui/src/settings_dialog.cpp [hunk 2]
```diff
@@ -412,14 +428,27 @@ void draw_settings_dialog(GuiState &gui, HostState &host) {
...
+        ImGui::TextColored(GUI_COLOR_TEXT_TITLE, "Upscaling");
+        ImGui::Spacing();
+        if (!host.io.title_id.empty())
+            ImGui::BeginDisabled();
+        if (ImGui::Combo("Resolution multiplier", &resolution_multiplier_idx, LIST_RES_MULTIPLIERS, IM_ARRAYSIZE(LIST_RES_MULTIPLIERS)))
+            config.resolution_multiplier = resolution_multiplier_idx + 1;
+        if (ImGui::IsItemHovered())
+            ImGui::SetTooltip("Enable upscaling for Vita3K. Experimental: games are not guaranteed to render properly at more than 1x");
+        if (!host.io.title_id.empty())
+            ImGui::EndDisabled();
```
Comment (Zangetsu38) [inline vita3k/gui/src/settings_dialog.cpp:451]:
```suggestion
        if (!host.io.title_id.empty()) {
            if (ImGui::Combo("Resolution multiplier", &resolution_multiplier_idx, LIST_RES_MULTIPLIERS, IM_ARRAYSIZE(LIST_RES_MULTIPLIERS)))
                config.resolution_multiplier = resolution_multiplier_idx + 1;
            if (ImGui::IsItemHovered())
                ImGui::SetTooltip("Enable upscaling for Vita3K. Experimental: games are not guaranteed to render properly at more than 1x");
            }
```

Comment (Macdu) [reply vita3k/gui/src/settings_dialog.cpp:451]: I think that making it disabled looks cleaner than the option literally disappearing while ingame.

Comment (Zangetsu38) [reply vita3k/gui/src/settings_dialog.cpp:451]: disable ? ah is no juste hide ?

Comment (Zangetsu38) [reply vita3k/gui/src/settings_dialog.cpp:451]: ok my bad lol

**Comment by Zangetsu38** [comment]
You can merge two commit, on this situation upsaling is added with all other code, no split it in this case :)

**Review by Zangetsu38 — APPROVED** [review]

**Pull request merged**